src/config.rs

```diff
@@ -1,64 +1,184 @@
-
-    fn limit(&mut self) -> &mut Self::Limit;
-}
+/// The default options for bincode serialization/deserialization.
+/// Implements OptionsExt to allow building configuration object for non-default settings.
+///
+/// ### Defaults
+/// By default bincode will use little-endian encoding for mult-byte integers, and will not
+/// limit the number of serialized/deserialized bytes.
+#[derive(Copy, Clone)]
+pub struct DefaultOptions(Infinite);
 
-pub(crate) trait OptionsExt: Options + Sized {
+/// A configuration builder trait whose options Bincode will use
+/// while serializing and deserializing.
+///
+/// ### Options
+/// Endianness: The endianness with which multi-byte integers will be read/written.  *default: little endian*
+/// Limit: The maximum number of bytes that will be read/written in a bincode serialize/deserialize. *default: unlimited*
+///
+/// ### Byte Limit Details
+/// The purpose of byte-limiting is to prevent Denial-Of-Service attacks whereby malicious attackers get bincode
+/// deserialization to crash your process by allocating too much memory or keeping a connection open for too long.
+///
+/// When a byte limit is set, bincode will return `Err` on any deserialization that goes over the limit, or any
+/// serialization that goes over the limit.
+/// Sets the byte limit to be unlimited.
+/// This is the default.
+pub trait OptionsExt: Options + Sized {
+    /// Sets the byte limit to be unlimited.
+    /// This is the default.
     fn with_no_limit(self) -> WithOtherLimit<Self, Infinite> {
         WithOtherLimit::new(self, Infinite)
     }
 
+    /// Sets the byte limit to `limit`.
     fn with_limit(self, limit: u64) -> WithOtherLimit<Self, Bounded> {
         WithOtherLimit::new(self, Bounded(limit))
     }
 
+    /// Sets the endianness to little-endian
+    /// This is the default.
     fn with_little_endian(self) -> WithOtherEndian<Self, LittleEndian> {
         WithOtherEndian::new(self)
     }
 
+    /// Sets the endianness to big-endian
     fn with_big_endian(self) -> WithOtherEndian<Self, BigEndian> {
         WithOtherEndian::new(self)
     }
 
+    /// Sets the endianness to the the machine-native endianness
     fn with_native_endian(self) -> WithOtherEndian<Self, NativeEndian> {
         WithOtherEndian::new(self)
     }
-}
 
-impl<'a, O: Options> Options for &'a mut O {
-    type Limit = O::Limit;
-    type Endian = O::Endian;
+    /// Serializes a serializable object into a `Vec` of bytes using this configuration
+    #[inline(always)]
+    fn serialize<S: ?Sized + serde::Serialize>(&self, t: &S) -> Result<Vec<u8>> {
+        ::internal::serialize(t, self.clone())
+    }
 
+    /// Returns the size that an object would be if serialized using Bincode with this configuration
     #[inline(always)]
-    fn limit(&mut self) -> &mut Self::Limit {
-        (*self).limit()
+    fn serialized_size<T: ?Sized + serde::Serialize>(&self, t: &T) -> Result<u64> {
+        ::internal::serialized_size(t, self.clone())
```

Comment:
I think it would be best if all these functions consumed `self` instead of taking a reference and calling `.clone()`. This gives the caller the option to call `.clone()` themselves or let rust simply consume the `Options`.

Currently the types are zero-sized so it won't matter in performance. 

Changing this to consume `self` is more in-line with the other (existing) methods.